src/Specular/Internal/Incremental/PriorityQueue.js

```diff
@@ -0,0 +1,73 @@
+// new :: forall a.
+//   EffectFn4
+//     (Optional Any)                 -- Optional.none
+//     (Field a Mutable Int)          -- Priority
+//     (Field a Mutable Boolean)      -- Is the entry present in this queue?
+//     (Field a Mutable (Optional a)) -- Next entry with the same priority
+//   (PQ a)
+exports.new = function(none, priorityField, presentField, nextField) {
+  return {
+    none: none,
+
+    // Property names of various properties of queue elements.
+    priorityField: priorityField,
+    presentField: presentField,
+    nextField: nextField,
+
+    // Array indexed by priority.
+    // Elements within each priority are stored in an intrusive linked list:
+    // - `this.none` is the end
+    // - `element[this.nextField]` is the next element.
+    // `priorityHeads[priority]` contains the head of this list (or `none` if empty).
+    priorityHeads: [],
+
+    // Total number of elements in the queue.
+    count: 0,
+
+    // TODO: we should track minPriority
+  };
+};
+
+// add :: forall a. EffectFn2 (PQ a) a Boolean
+exports.add = function(pq, node) {
```

Comment:
Add an error/warning if a certain size is reached?

Reply:
I wouldn't worry about size, rather max priority.

Would a `console.warn` be enough?

Reply:
Yup.